apollo-federation/src/operation/mod.rs

```diff
@@ -2503,29 +2534,47 @@ impl SelectionSet {
     /// Inserts a `Selection` into the inner map. Should a selection with the same key already
     /// exist in the map, the existing selection and the given selection are merged, replacing the
     /// existing selection while keeping the same insertion index.
-    fn add_selection(&mut self, selection: Selection) -> Result<(), FederationError> {
-        let selections = Arc::make_mut(&mut self.selections);
-
-        let key = selection.key();
-        match selections.remove(&key) {
-            Some((index, existing_selection)) => {
-                let to_merge = [existing_selection, selection];
-                // `existing_selection` and `selection` both have the same selection key,
-                // so the merged selection will also have the same selection key.
-                let selection = SelectionSet::make_selection(
-                    &self.schema,
-                    &self.type_position,
-                    to_merge.iter(),
-                    /*named_fragments*/ &Default::default(),
-                )?;
-                selections.insert_at(index, selection);
-            }
-            None => {
-                selections.insert(selection);
-            }
-        }
+    pub(crate) fn add_selection(&mut self, selection: &Selection) -> Result<(), FederationError> {
+        debug_assert_eq!(
+            &self.schema,
+            selection.schema(),
+            "In order to add selection it needs to point to the same schema"
+        );
+        self.merge_selections_into(std::iter::once(selection))
+    }
 
-        Ok(())
+    /// Inserts a `SelectionSet` into the inner map. Should any sub selection with the same key already
+    /// exist in the map, the existing selection and the given selection are merged, replacing the
+    /// existing selection while keeping the same insertion index.
+    pub(crate) fn add_selection_set(
+        &mut self,
+        selection_set: &SelectionSet,
+    ) -> Result<(), FederationError> {
+        debug_assert_eq!(
+            self.schema, selection_set.schema,
+            "In order to add selection set it needs to point to the same schema"
+        );
+        debug_assert_eq!(
+            self.type_position, selection_set.type_position,
+            "In order to add selection set it needs to point to the same type position"
+        );
+        self.merge_into(std::iter::once(selection_set))
+    }
+
+    /// Rebase given `SelectionSet` on self and then inserts it into the inner map. Should any sub
+    /// selection with the same key already exist in the map, the existing selection and the given
+    /// selection are merged, replacing the existing selection while keeping the same insertion index.
+    pub(crate) fn add_selection_set_with_rebase(
```

Comment:
any suggestions for a name? technically it is `rebase_and_add_selection_set` but thought it might be easier to discover if they all start with same prefix